fdbrpc/AsyncFileChaos.actor.h

```diff
@@ -26,94 +26,103 @@
 #include "flow/ActorCollection.h"
 #include "flow/actorcompiler.h"
 
-
-//template <class AsyncFileType>
+// template <class AsyncFileType>
 class AsyncFileChaos final : public IAsyncFile, public ReferenceCounted<AsyncFileChaos> {
 private:
 	Reference<IAsyncFile> file;
+	Arena arena;
+
 public:
 	explicit AsyncFileChaos(Reference<IAsyncFile> file) : file(file) {}
 
 	void addref() override { ReferenceCounted<AsyncFileChaos>::addref(); }
 	void delref() override { ReferenceCounted<AsyncFileChaos>::delref(); }
 
-	uint8_t toggleNthBit(uint8_t b, uint8_t n) {
-		auto singleBitMask = uint8_t(1) << (n);
-		return b ^ singleBitMask;
-	}
+	static double getDelay() {
+		double delayFor = 0.0;
+		auto res = g_network->global(INetwork::enDiskFailureInjector);
+		if (res) {
+			DiskFailureInjector* delayInjector = static_cast<DiskFailureInjector*>(res);
+			delayFor = delayInjector->getDiskDelay();
 
-	void flipBits(void* data, int length, double percentBitFlips) {
-		auto toFlip = int(float(length*8) * percentBitFlips / 100);
-		TraceEvent("AsyncFileFlipBits").detail("ToFlip", toFlip);
-		for (auto i = 0; i < toFlip; i++) {
-			auto byteOffset = deterministicRandom()->randomInt64(0, length);
-			auto bitOffset = uint8_t(deterministicRandom()->randomInt(0, 8));
-			((uint8_t *)data)[byteOffset] = toggleNthBit(((uint8_t *)data)[byteOffset], bitOffset);
+			// increment the metric for disk delays
+			if (delayFor > 0.0) {
+				auto res = g_network->global(INetwork::enChaosMetrics);
+				if (res) {
+					ChaosMetrics* chaosMetrics = static_cast<ChaosMetrics*>(res);
+					chaosMetrics->diskDelays++;
+				}
+			}
 		}
+		return delayFor;
 	}
 
 	Future<int> read(void* data, int length, int64_t offset) override {
-		double delay = 0.0;
-		auto res = g_network->global(INetwork::enDiskFailureInjector);
-		if (res)
-			delay = static_cast<DiskFailureInjector*>(res)->getDiskDelay();
-		TraceEvent("AsyncFileChaosRead").detail("ThrottleDelay", delay);
-		return delayed(file->read(data, length, offset), delay);
+		double diskDelay = getDelay();
+
+		// Wait for diskDelay before submitting the I/O
```

Comment:
If the delay is 0, just return the read future directly.  `delay(0)` is far from a no-op.

Reply:
Done